app/controllers/api/providers_controller.rb

```diff
@@ -85,6 +85,15 @@ def import_vm_resource(type, id = nil, data = {})
       end
     end
 
+    def options
+      req_class_name = @req.to_hash[:fullpath]
+                           .split('?').last.split('&')                        # '?' for query, '&' connects parameters
+                           .select { |param| param.starts_with?("ems_type") } # 'parameters with "ems_type" name
+                           .first.split('=').last                             # choosing only the first occuracne.
+      ems_class = ExtManagementSystem.descendants.select { |ems| ems.ems_type == req_class_name }.first
```

Comment:
missed this earlier, but you should be able to get the ems type from just params[:ems_type]. Then you can actually use `ExtManagementSystem.descendant_get(params[:ems_type])` here. That will return the descendant. An example of its usage can be found https://github.com/ManageIQ/manageiq-api/blob/master/lib/services/api/authentication_service.rb#L4